File: src/main/java/com/faendir/acra/rest/RestConfiguration.java
/*
 * (C) Copyright 2018 Lukas Morawietz (https://github.com/F43nd1r)
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package com.faendir.acra.rest;

import com.faendir.acra.rest.multipart.Rfc1341MultipartResolver;
import com.github.ziplet.filter.compression.CompressingFilter;
import org.springframework.context.annotation.Bean;
import org.springframework.context.annotation.Configuration;
import org.springframework.core.Ordered;
import org.springframework.core.annotation.Order;
import org.springframework.lang.NonNull;
import org.springframework.web.multipart.MultipartResolver;

import javax.servlet.Filter;

/**
 * @author lukas
 * @since 16.05.18
 */
@Configuration
public class RestConfiguration {
    @NonNull
    @Bean
    public static MultipartResolver multiPartResolver() {
        return new Rfc1341MultipartResolver();
    }

    @Bean
    @Order(Ordered.HIGHEST_PRECEDENCE)
    public Filter gzipFilter() {
        return new CompressingFilter();
    }
}

File: acrarium/src/main/kotlin/com/faendir/acra/rest/RestConfiguration.kt
package com.faendir.acra.rest

import com.github.ziplet.filter.compression.CompressingFilter
import org.springframework.context.annotation.Bean
import org.springframework.context.annotation.Configuration
import org.springframework.core.Ordered
import org.springframework.core.annotation.Order
import org.springframework.web.multipart.MultipartHttpServletRequest
import org.springframework.web.multipart.MultipartResolver
import org.springframework.web.multipart.commons.CommonsMultipartResolver
import org.springframework.web.multipart.support.StandardServletMultipartResolver
import javax.servlet.Filter
import javax.servlet.http.HttpServletRequest

@Configuration
class RestConfiguration {
    @Bean
    @Order(Ordered.HIGHEST_PRECEDENCE)
    fun gzipFilter(): Filter = CompressingFilter()

    @Bean
    fun multipartResolver(): MultipartResolver = SwitchingMultipartResolver()
}

class SwitchingMultipartResolver : MultipartResolver {
    private val commonsResolver = CommonsMultipartResolver()
    private val standardResolver = StandardServletMultipartResolver()
    override fun isMultipart(request: HttpServletRequest): Boolean {
        return request.resolver().isMultipart(request)
    }

    override fun resolveMultipart(request: HttpServletRequest): MultipartHttpServletRequest {
        return request.resolver().resolveMultipart(request)
    }

    override fun cleanupMultipart(request: MultipartHttpServletRequest) {
        return request.resolver().cleanupMultipart(request)
    }

    private fun HttpServletRequest.resolver() = if (this.servletPath.startsWith("/VAADIN")) standardResolver else commonsResolver
}